sklearn/linear_model/ridge.py

```diff
@@ -492,24 +540,33 @@ def fit(self, X, y, sample_weight=None):
                 np.atleast_1d(sample_weight).ndim > 1):
             raise ValueError("Sample weights must be 1D array or scalar")
 
+        # when X is sparse we only remove offset from y
         X, y, X_offset, y_offset, X_scale = self._preprocess_data(
             X, y, self.fit_intercept, self.normalize, self.copy_X,
-            sample_weight=sample_weight)
+            sample_weight=sample_weight, return_mean=True)
 
         # temporary fix for fitting the intercept with sparse data using 'sag'
-        if sparse.issparse(X) and self.fit_intercept:
-            self.coef_, self.n_iter_, self.intercept_ = ridge_regression(
+        if (sparse.issparse(X) and self.fit_intercept and
+           self.solver != 'sparse_cg'):
```

Comment:
can you add here:

`assert self.solver == 'sag'`

to make it clear that we ware in the sag branch?

Reply:
actually, it could be any solver except `sparse_cg` because `ridge_regression` will automatically switch to `sag` and raise a warning 🤯 This logic is far from optimal, but it's not really in the scope of this PR.